### ISSUE TITLE
Composer require causes issues in cache clearing

### ISSUE THREAD

**Issue by arendjantetteroo** [iwink/gitlab-webhook-bundle#10]
**Describe the bug**
Just installing this bundle (with some sf6 allows in my fork) gives me the following error:

```
Symfony operations: 1 recipe (c20eafe62fba639dac34dbf0117e99d1)
  - Configuring iwink/gitlab-webhook-bundle (>=dev-symfony6): From auto-generated recipe
Executing script cache:clear [KO]
 [KO]
Script cache:clear returned with error code 1
!!  
!!   // Clearing the cache for the dev environment with debug true                  
!!  
!!  
!!  In FileLoader.php line 178:
!!                                                                                 
!!    You have requested a non-existent parameter "secret" in /srv/app/config/rou  
!!    tes/../../src/Kernel.php (which is being imported from "/srv/app/config/rou  
!!    tes/annotations.yaml"). Make sure there is a loader supporting the "annotat  
!!    ion" type.                                                                   
!!                                                                                 
!!  
!!  In App_KernelDevDebugContainer.php line 3024:
!!                                                           
!!    You have requested a non-existent parameter "secret".  
!!                                                           
!!  
!!  cache:clear [--no-warmup] [--no-optional-warmers]
```

Looking at the code that is responsible: 

````
In FileLoader.php line 178:
                                                                                                                                                                                                                                               
  [Symfony\Component\Config\Exception\LoaderLoadException]                                                                                                                                                                                     
  You have requested a non-existent parameter "secret" in /srv/app/config/routes/../../src/Kernel.php (which is being imported from "/srv/app/config/routes/annotations.yaml"). Make sure there is a loader supporting the "annotation" type.  
                                                                                                                                                                                                                                               

Exception trace:
  at /srv/app/vendor/symfony/config/Loader/FileLoader.php:178
 Symfony\Component\Config\Loader\FileLoader->doImport() at /srv/app/vendor/symfony/config/Loader/FileLoader.php:94
 Symfony\Component\Config\Loader\FileLoader->import() at /srv/app/vendor/symfony/routing/Loader/YamlFileLoader.php:208
 Symfony\Component\Routing\Loader\YamlFileLoader->parseImport() at /srv/app/vendor/symfony/routing/Loader/YamlFileLoader.php:99
 Symfony\Component\Routing\Loader\YamlFileLoader->load() at /srv/app/vendor/symfony/config/Loader/FileLoader.php:163
 Symfony\Component\Config\Loader\FileLoader->doImport() at /srv/app/vendor/symfony/config/Loader/FileLoader.php:83
 Symfony\Component\Config\Loader\FileLoader->import() at /srv/app/vendor/symfony/routing/Loader/Configurator/RoutingConfigurator.php:45
 Symfony\Component\Routing\Loader\Configurator\RoutingConfigurator->import() at /srv/app/vendor/symfony/framework-bundle/Kernel/MicroKernelTrait.php:78
 App\Kernel->configureRoutes() at /srv/app/vendor/symfony/framework-bundle/Kernel/MicroKernelTrait.php:202
 App\Kernel->loadRoutes() at /srv/app/vendor/symfony/routing/Loader/ObjectLoader.php:55
 Symfony\Component\Routing\Loader\ObjectLoader->load() at /srv/app/vendor/symfony/config/Loader/DelegatingLoader.php:37
 Symfony\Component\Config\Loader\DelegatingLoader->load() at /srv/app/vendor/symfony/framework-bundle/Routing/DelegatingLoader.php:67
 Symfony\Bundle\FrameworkBundle\Routing\DelegatingLoader->load() at /srv/app/vendor/symfony/framework-bundle/Routing/Router.php:65
 Symfony\Bundle\FrameworkBundle\Routing\Router->getRouteCollection() at /srv/app/vendor/symfony/routing/Router.php:317
 Symfony\Component\Routing\Router->getMatcherDumperInstance() at /srv/app/vendor/symfony/routing/Router.php:255
 Symfony\Component\Routing\Router->Symfony\Component\Routing\{closure}() at /srv/app/vendor/symfony/config/ResourceCheckerConfigCacheFactory.php:36
 Symfony\Component\Config\ResourceCheckerConfigCacheFactory->cache() at /srv/app/vendor/symfony/routing/Router.php:253
 Symfony\Component\Routing\Router->getMatcher() at /srv/app/vendor/symfony/framework-bundle/Routing/Router.php:92
 Symfony\Bundle\FrameworkBundle\Routing\Router->warmUp() at /srv/app/vendor/symfony/framework-bundle/CacheWarmer/RouterCacheWarmer.php:42
 Symfony\Bundle\FrameworkBundle\CacheWarmer\RouterCacheWarmer->warmUp() at /srv/app/vendor/symfony/http-kernel/CacheWarmer/CacheWarmerAggregate.php:96
 Symfony\Component\HttpKernel\CacheWarmer\CacheWarmerAggregate->warmUp() at /srv/app/vendor/symfony/framework-bundle/Command/CacheClearCommand.php:236
 Symfony\Bundle\FrameworkBundle\Command\CacheClearCommand->warmup() at /srv/app/vendor/symfony/framework-bundle/Command/CacheClearCommand.php:148
 Symfony\Bundle\FrameworkBundle\Command\CacheClearCommand->execute() at /srv/app/vendor/symfony/console/Command/Command.php:312
 Symfony\Component\Console\Command\Command->run() at /srv/app/vendor/symfony/console/Application.php:1040
 Symfony\Component\Console\Application->doRunCommand() at /srv/app/vendor/symfony/framework-bundle/Console/Application.php:88
 Symfony\Bundle\FrameworkBundle\Console\Application->doRunCommand() at /srv/app/vendor/symfony/console/Application.php:314
 Symfony\Component\Console\Application->doRun() at /srv/app/vendor/symfony/framework-bundle/Console/Application.php:77
 Symfony\Bundle\FrameworkBundle\Console\Application->doRun() at /srv/app/vendor/symfony/console/Application.php:168
 Symfony\Component\Console\Application->run() at /srv/app/vendor/symfony/runtime/Runner/Symfony/ConsoleApplicationRunner.php:54
 Symfony\Component\Runtime\Runner\Symfony\ConsoleApplicationRunner->run() at /srv/app/vendor/autoload_runtime.php:29
 require_once() at /srv/app/bin/console:11

In App_KernelDevDebugContainer.php line 3024:
                                                                                
  [Symfony\Component\DependencyInjection\Exception\ParameterNotFoundException]  
  You have requested a non-existent parameter "secret".                         
                                                                                

Exception trace:
  at /srv/app/var/cache/dev/ContainerO8uiHFi/App_KernelDevDebugContainer.php:3024
 ContainerO8uiHFi\App_KernelDevDebugContainer->getParameter() at /srv/app/vendor/symfony/dependency-injection/ParameterBag/ContainerBag.php:35
 Symfony\Component\DependencyInjection\ParameterBag\ContainerBag->get() at /srv/app/vendor/symfony/dependency-injection/ParameterBag/ParameterBag.php:190
 Symfony\Component\DependencyInjection\ParameterBag\ParameterBag->resolveString() at /srv/app/vendor/symfony/dependency-injection/ParameterBag/ParameterBag.php:164
 Symfony\Component\DependencyInjection\ParameterBag\ParameterBag->resolveValue() at /srv/app/vendor/iwink/gitlab-webhook-bundle/src/Annotation/Reflection/ParameterAwareTrait.php:37
 Iwink\GitLabWebhookBundle\Annotation\Reflection\ReflectionMethod->Iwink\GitLabWebhookBundle\Annotation\Reflection\{closure}() at n/a:n/a
 preg_replace_callback() at /srv/app/vendor/iwink/gitlab-webhook-bundle/src/Annotation/Reflection/ParameterAwareTrait.php:34
 Iwink\GitLabWebhookBundle\Annotation\Reflection\ReflectionMethod->getDocComment() at /srv/app/vendor/doctrine/annotations/lib/Doctrine/Common/Annotations/AnnotationReader.php:206
 Doctrine\Common\Annotations\AnnotationReader->getMethodAnnotations() at /srv/app/vendor/iwink/gitlab-webhook-bundle/src/Annotation/ParameterAwareReader.php:55
 Iwink\GitLabWebhookBundle\Annotation\ParameterAwareReader->getMethodAnnotations() at /srv/app/vendor/doctrine/annotations/lib/Doctrine/Common/Annotations/PsrCachedReader.php:155
 Doctrine\Common\Annotations\PsrCachedReader->fetchFromCache() at /srv/app/vendor/doctrine/annotations/lib/Doctrine/Common/Annotations/PsrCachedReader.php:119
 Doctrine\Common\Annotations\PsrCachedReader->getMethodAnnotations() at /srv/app/vendor/symfony/routing/Loader/AnnotationClassLoader.php:364
 Symfony\Component\Routing\Loader\AnnotationClassLoader->getAnnotations() at /srv/app/vendor/symfony/routing/Loader/AnnotationClassLoader.php:129
 Symfony\Component\Routing\Loader\AnnotationClassLoader->load() at /srv/app/vendor/symfony/routing/Loader/AnnotationFileLoader.php:57
 Symfony\Component\Routing\Loader\AnnotationFileLoader->load() at /srv/app/vendor/symfony/routing/Loader/AnnotationDirectoryLoader.php:31
 Symfony\Component\Routing\Loader\AnnotationDirectoryLoader->load() at /srv/app/vendor/symfony/config/Loader/FileLoader.php:163
 Symfony\Component\Config\Loader\FileLoader->doImport() at /srv/app/vendor/symfony/config/Loader/FileLoader.php:94
 Symfony\Component\Config\Loader\FileLoader->import() at /srv/app/vendor/symfony/routing/Loader/YamlFileLoader.php:208
 Symfony\Component\Routing\Loader\YamlFileLoader->parseImport() at /srv/app/vendor/symfony/routing/Loader/YamlFileLoader.php:99
 Symfony\Component\Routing\Loader\YamlFileLoader->load() at /srv/app/vendor/symfony/config/Loader/FileLoader.php:163
 Symfony\Component\Config\Loader\FileLoader->doImport() at /srv/app/vendor/symfony/config/Loader/FileLoader.php:83
 Symfony\Component\Config\Loader\FileLoader->import() at /srv/app/vendor/symfony/routing/Loader/Configurator/RoutingConfigurator.php:45
 Symfony\Component\Routing\Loader\Configurator\RoutingConfigurator->import() at /srv/app/vendor/symfony/framework-bundle/Kernel/MicroKernelTrait.php:78
 App\Kernel->configureRoutes() at /srv/app/vendor/symfony/framework-bundle/Kernel/MicroKernelTrait.php:202
 App\Kernel->loadRoutes() at /srv/app/vendor/symfony/routing/Loader/ObjectLoader.php:55
 Symfony\Component\Routing\Loader\ObjectLoader->load() at /srv/app/vendor/symfony/config/Loader/DelegatingLoader.php:37
 Symfony\Component\Config\Loader\DelegatingLoader->load() at /srv/app/vendor/symfony/framework-bundle/Routing/DelegatingLoader.php:67
 Symfony\Bundle\FrameworkBundle\Routing\DelegatingLoader->load() at /srv/app/vendor/symfony/framework-bundle/Routing/Router.php:65
 Symfony\Bundle\FrameworkBundle\Routing\Router->getRouteCollection() at /srv/app/vendor/symfony/routing/Router.php:317
 Symfony\Component\Routing\Router->getMatcherDumperInstance() at /srv/app/vendor/symfony/routing/Router.php:255
 Symfony\Component\Routing\Router->Symfony\Component\Routing\{closure}() at /srv/app/vendor/symfony/config/ResourceCheckerConfigCacheFactory.php:36
 Symfony\Component\Config\ResourceCheckerConfigCacheFactory->cache() at /srv/app/vendor/symfony/routing/Router.php:253
 Symfony\Component\Routing\Router->getMatcher() at /srv/app/vendor/symfony/framework-bundle/Routing/Router.php:92
 Symfony\Bundle\FrameworkBundle\Routing\Router->warmUp() at /srv/app/vendor/symfony/framework-bundle/CacheWarmer/RouterCacheWarmer.php:42
 Symfony\Bundle\FrameworkBundle\CacheWarmer\RouterCacheWarmer->warmUp() at /srv/app/vendor/symfony/http-kernel/CacheWarmer/CacheWarmerAggregate.php:96
 Symfony\Component\HttpKernel\CacheWarmer\CacheWarmerAggregate->warmUp() at /srv/app/vendor/symfony/framework-bundle/Command/CacheClearCommand.php:236
 Symfony\Bundle\FrameworkBundle\Command\CacheClearCommand->warmup() at /srv/app/vendor/symfony/framework-bundle/Command/CacheClearCommand.php:148
 Symfony\Bundle\FrameworkBundle\Command\CacheClearCommand->execute() at /srv/app/vendor/symfony/console/Command/Command.php:312
 Symfony\Component\Console\Command\Command->run() at /srv/app/vendor/symfony/console/Application.php:1040
 Symfony\Component\Console\Application->doRunCommand() at /srv/app/vendor/symfony/framework-bundle/Console/Application.php:88
 Symfony\Bundle\FrameworkBundle\Console\Application->doRunCommand() at /srv/app/vendor/symfony/console/Application.php:314
 Symfony\Component\Console\Application->doRun() at /srv/app/vendor/symfony/framework-bundle/Console/Application.php:77
 Symfony\Bundle\FrameworkBundle\Console\Application->doRun() at /srv/app/vendor/symfony/console/Application.php:168
 Symfony\Component\Console\Application->run() at /srv/app/vendor/symfony/runtime/Runner/Symfony/ConsoleApplicationRunner.php:54
 Symfony\Component\Runtime\Runner\Symfony\ConsoleApplicationRunner->run() at /srv/app/vendor/autoload_runtime.php:29
 require_once() at /srv/app/bin/console:11

cache:clear [--no-warmup] [--no-optional-warmers]
`
```

It seems the parameterAwareTrait is just reading all docblocks for my whole codebase, instead of just docblocks that are relevant for the Webhook annotation, that i havent' even added yet. 

Is this on purpose?

**Comment by arendjantetteroo** [comment]
Last couple of lines on line 34 of the ParameterAwareTrait:

```
`
ParameterAwareTrait.php on line 34:
"     * We don't use the PHP php_strip_whitespace() function"
ParameterAwareTrait.php on line 34:
"     * as we want the content to be readable and well-formatted."
ParameterAwareTrait.php on line 34:
"     */"
ParameterAwareTrait.php on line 34:
"/**"
ParameterAwareTrait.php on line 34:
"     * Configures the container."
ParameterAwareTrait.php on line 34:
"     *"
ParameterAwareTrait.php on line 34:
"     * You can register extensions:"
ParameterAwareTrait.php on line 34:
"     *"
ParameterAwareTrait.php on line 34:
"     *     $container->extension('framework', ["
ParameterAwareTrait.php on line 34:
"     *         'secret' => '%secret%'"
/srv/app # 
```

`

**Comment by arendjantetteroo** [comment]
From the documentation i see that you are supposed to be able to do this:

` `@Webhook("pipeline", tokens={"%gitlab.secret_token%"})``

If that's possible, why not just resolve that parameter when you use the webhook annotation? 
If you see Webhook::getTokens() containing the % strings, just look those tokens up in the GitlabWebhookSubscriber? 
Then you don't need a parameterAwareReader?

**Comment by doekenorg** [comment]
It's not so much that we are reading every class; but Symfony does this with the `annotations.reader` service. Our bundle expands on this, by replacing the reader with a custom version. But this could very well be a decorator. It looks like we didn't take into account other classes that had this exact definition. 

We'll take a look at it. I think you're right about only checking out our own annotations, instead of all of them.

**Comment by arendjantetteroo** [comment]
@doekenorg As i needed to disable the reader to get it to work, i've added a proposal branch for how this could work. 

https://github.com/iwink/gitlab-webhook-bundle/pull/11

I assumed that you only want to lookup the tokens in the parameterBag. If so (and it's enough for my usecase) i moved that part to the GitlabWebhookSubscriber and removed all the reader logic. 
As this subscriber logic is only called once a webhook annotation has been found, there should not be a performance penalty. 

It currently includes the symfony6 and doctrine/annotations2 allow in composer.json. 

I did not update the unit tests yet before knowing if this is an acceptable approach. 
I do have a couple integration tests on my own project that work with these changes.

**Comment by doekenorg** [comment]
Yeah I was looking a simple solution for now; and it looks like doing a `has` check on the parameter, and just returning the entire preg_match if it doesn't exist works as well. 

```php
function (array $matches): string {
    // If the parameter doesn't exist; don't replace it.
    if (!$this->parameterBag->has($matches[0])) {
        return $matches[0];
    }

    $parameter = $this->parameterBag->resolveValue($matches[0]);

    return is_array($parameter) ? sprintf('{"%s"}', implode('","', $parameter)) : $parameter;
```

I inherited this project from my ex-colleagues, so I have to play some catch up. And today is my day off; so I'm not pursuing this further today. But I'll make sure this gets some attention next week. Thanks again for your testing and time. Really appreciated.

**Comment by arendjantetteroo** [comment]
@doekenorg no rush, i can now continue with my work of getting the gitlab events into my system :) 
Enjoy your weekend!

**Comment by arendjantetteroo** [comment]
Just to inform you, I put my branch into production and it works great.

**Comment by doekenorg** [comment]
@arendjantetteroo This issue is resolved in `v1.2.2`. Although your solution is better, we'll plan to take that route later on in a new release. For now it's a bug for all users, including Symfony < 6, so I took the pragmatic route for today.
